src/main/webapp/dev/js/common/instructor.js

```diff
@@ -405,12 +419,17 @@ function bindSessionDeleteAllLinks() {
         event.preventDefault();
 
         const $clickedLink = $(event.currentTarget);
-        const messageText = 'Are you sure you want to permanently delete all feedback sessions?';
+        const softDeletedCourseIds = $('[id^="softdeletedcourseid"]').toArray()
```

Comment:
The course id and session name are data that belong together. Consider using an array of objects to associate the data.

For example,
```Javascript
const sessions = [
    {
        courseId: '1',
        sessionName: 'Session 1',
    },
    {
        courseId: '2',
        sessionName: 'Session 2',
    },
];
```

Reply:
Shall I map it directly to an array of strings (i.e. "courseId: sessionName") since we want to use a single method (as stated below) for generating HTML for both courses & sessions?

Reply:
Yes I agree. But in this case is it really necessary for us to keep the structure as we will only use it once (for the purpose of generating HTML) inside this method? So far I still cannot think of a way to do the abstraction if we keep the structure.

Reply:
It's not about keeping the structure even if we only use it once. It's about having a structure so we can easily see how the data is represented.

The structure is already roughly shown when we store them into `data-*` attributes so it is possible to implement without translating the `data-*` attributes into a `Session` data structure.

It is just going to depend on how you write the code and how we can see the relationships clearly.

> So far I still cannot think of a way to do the abstraction if we keep the structure.

There is no need to think too hard about this, it just simply means we first extract the data from `data-*` attributes and map it to presentation ( `${courseId}: ${sessionName}`). Whether that extra step to translate to a structure or to skip to presentation directly will depend on the understandability of the code I have mentioned.

Reply:
> There is no need to think too hard about this, it just simply means we first extract the data from data-* attributes as a data structure before mapping it to presentation ( ${courseId}: ${sessionName}). Whether that extra step to translate to a structure or to skip to presentation directly will depend on the understandability of the code I have mentioned.

Ehh...I'm not quite sure what you mean but are you suggesting something like this?
```javascript
const softDeletedSessions = $('[id^="softdeletedcourseid"]').toArray()
                .map(softDeletedSessionIdNamePair => ({ courseId: $(softDeletedSessionIdNamePair).data('softDeletedCourseId'), sessionsName: $(softDeletedSessionIdNamePair).data('softDeletedSessionName') }))
                .map(softDeletedSession => `${softDeletedSession.courseId}: ${softDeletedSession.sessionsName}`);
```

Reply:
Yes but if you write it in this way mapping it to a data structure will be rather redundant.

However, notice that `softDeletedSessions` is no longer a list of sessions but rather a list of one possible string representation of soft-deleted sessions.

So whether you rename `softDeletedSessions` to something else or whether you are going to first map it to a data structure first is going to depend on the approach you take.

Reply:
Assuming that we want to keep this function as is:
```javascript
/**
 * Returns the HTML required to render the soft deleted courses/sessions as an unordered list.
 */
function generateUnorderedListHtml(entityRows) {
    return `\
<ul>
${entityRows
            .map(entityRow => `<li>${entityRow}</li>`)
            .join('')}
</ul>`;
}
```

So in this case I can just extract the data and map it to a list of string representation of sessions?

Any suggestion on the attribute name (i.e. `softDeletedSessionsInfo`)?

Reply:
`entityRows` -> `items`. The reason is `entityRows` is a confusing name with no clear meaning and even if we document it, it isn't very ideal. It is also an unnecessary restriction on what the method can do and do not clearly convey the method intent.

What `generateUnorderedListHtml` is simply taking a list of items and then rendering it as a simple unordered list.

> Any suggestion on the attribute name (i.e. softDeletedSessionsInfo)?

We can observe the challenge in naming here, `softDeletedSessionsInfo` is too vague. `formattedSessionsAsCourseIdSessionPair`is seemingly appropriate but it is too verbose. Alternatively, extract a method that does the formatting and we can simply name the variable as something like `sessionItems`.

Reply:
```javascript
const softDeletedSessions = $('[id^="softdeletedcourseid"]').toArray()
                .map(sessionItems => ({ courseId: $(sessionItems).data('softDeletedCourseId'), sessionsName: $(sessionItems).data('softDeletedSessionName') }))
                .map(sessionItems => `${sessionItems.courseId}: ${sessionItems.sessionsName}`);
```
Like this?

Reply:
I don't understand. Isn't this the same thing you posted previously, https://github.com/TEAMMATES/teammates/pull/9039#discussion_r210868297?

Reply:
...Yea I also feel the same way :-p . But isn't that what you mean (for the renaming)?

Reply:
I think you misunderstood what I said, going by the conversation it should look something like this:

```Javascript
const softDeletedSessionItems = $('[id^="softdeletedcourseid"]').toArray()
                .map(sessionItems => ({ courseId: $(sessionItems).data('softDeletedCourseId'), sessionsName: $(sessionItems).data('softDeletedSessionName') }))
                .map(sessionItems => `${sessionItems.courseId}: ${sessionItems.sessionsName}`);
```

Reply:
Ohoh get it 👌